Author: Rene Kita <mail@rkta.de>
Date: Wed Oct 25 08:56:00 CEST 2023
Subject: Refactor URL history handling

loadHistory and saveHistory both are only used for the URL history which
is a global variable. Do not pass pointers, but access this global
directly. This has the advantage that the current history is updated
when saveUrlHistory (former saveHistory) is called.

This is in preparation for the next patch.
---
 history.c |   89 +++++++++++++++++++++++++++++++++++---------------------------
 history.h |    4 +-
 main.c    |    4 +-
 3 files changed, 55 insertions(+), 42 deletions(-)

--- a/history.c
+++ b/history.c
@@ -2,19 +2,6 @@
 #include "fm.h"
 
 #ifdef USE_HISTORY
-/* Merge entries from their history into ours */
-static int
-mergeHistory(Hist *ours, Hist *theirs)
-{
-    HistItem *item;
-
-    for (item = theirs->list->first; item; item = item->next)
-	if (!getHashHist(ours, item->ptr))
-	    pushHist(ours, (char *)item->ptr);
-
-    return 0;
-}
-
 Buffer *
 historyBuffer(Hist *hist)
 {
@@ -44,14 +31,49 @@ historyBuffer(Hist *hist)
     return loadHTMLString(src);
 }
 
+/*
+ * If HISTORY_FILE was written to after we last read it, read it again and
+ * merge our URLHist into it.
+ */
+static int
+syncUrlHistory(void)
+{
+    Hist *hist;
+    HistItem *item;
+    struct stat st;
+
+    if (stat(rcFile(HISTORY_FILE), &st))
+	goto err;
+
+    hist = newHist();
+    if (loadUrlHistory())
+	goto err;
+
+    if (hist->mtime == (long long)st.st_mtime)
+	return 0;
+
+    /* Merge */
+    for (item = URLHist->list->first; item; item = item->next)
+	if (!getHashHist(hist, item->ptr))
+	    pushHist(hist, (char *)item->ptr);
+
+    URLHist = hist;
+
+    return 0;
+
+err:
+    disp_err_message("Can't sync URL history", FALSE);
+    return 1;
+}
+
 int
-loadHistory(Hist *hist)
+loadUrlHistory(void)
 {
     FILE *f;
     Str line;
     struct stat st;
 
-    if (hist == NULL)
+    if (URLHist == NULL)
 	return 1;
     if ((f = fopen(rcFile(HISTORY_FILE), "rt")) == NULL)
 	return 1;
@@ -60,7 +82,7 @@ loadHistory(Hist *hist)
 	fclose(f);
 	return 1;
     }
-    hist->mtime = (long long)st.st_mtime;
+    URLHist->mtime = (long long)st.st_mtime;
 
     while (!feof(f)) {
 	line = Strfgets(f);
@@ -69,47 +91,38 @@ loadHistory(Hist *hist)
 	Strremovetrailingspaces(line);
 	if (line->length == 0)
 	    continue;
-	pushHist(hist, url_quote(line->ptr));
+	pushHist(URLHist, url_quote(line->ptr));
     }
     fclose(f);
     return 0;
 }
 
 void
-saveHistory(Hist *hist, size_t size)
+saveUrlHistory(void)
 {
     FILE *f;
-    Hist *fhist;
     HistItem *item;
-    char *histf;
     char *tmpf;
-    int rename_ret;
-    struct stat st;
 
-    if (hist == NULL || hist->list == NULL)
+    if (URLHist == NULL || URLHist->list == NULL)
 	return;
 
-    histf = rcFile(HISTORY_FILE);
-    if (!stat(histf, &st) && hist->mtime != (long long)st.st_mtime) {
-	fhist = newHist();
-	if (loadHistory(fhist) || mergeHistory(fhist, hist))
-	    disp_err_message("Can't merge history", FALSE);
-	else
-	    hist = fhist;
-    }
+    syncUrlHistory();
 
     tmpf = tmpfname(TMPF_HIST, NULL)->ptr;
     if ((f = fopen(tmpf, "w")) == NULL)
 	goto fail;
-    for (item = hist->list->first; item && hist->list->nitem > size;
-	 item = item->next)
-	size++;
+
+    /* Respect URLHistSize before saving */
+    for (item = URLHist->list->first;
+	 item && URLHist->list->nitem > URLHistSize;
+	 item = item->next);
+
     for (; item; item = item->next)
 	fprintf(f, "%s\n", (char *)item->ptr);
     if (fclose(f) == EOF)
 	goto fail;
-    rename_ret = rename(tmpf, rcFile(HISTORY_FILE));
-    if (rename_ret != 0)
+    if (rename(tmpf, rcFile(HISTORY_FILE)))
 	goto fail;
 
     return;
@@ -121,8 +134,8 @@ fail:
 #endif				/* USE_HISTORY */
 
 /*
- * The following functions are used for internal stuff, we need them regardless
- * if history is used or not.
+ * The following functions are used for internal stuff, we need them
+ * regardless if history is used or not.
  */
 
 Hist *
--- a/history.h
+++ b/history.h
@@ -30,8 +30,8 @@ extern char *nextHist(Hist *hist);
 extern char *prevHist(Hist *hist);
 
 #ifdef USE_HISTORY
-extern int loadHistory(Hist *hist);
-extern void saveHistory(Hist *hist, size_t size);
+extern int loadUrlHistory(void);
+extern void saveUrlHistory(void);
 extern void ldHist(void);
 #else				/* not USE_HISTORY */
 #define ldHist nulcmd
--- a/main.c
+++ b/main.c
@@ -910,7 +910,7 @@ main(int argc, char **argv)
 #endif				/* USE_COOKIE */
 #ifdef USE_HISTORY
     if (UseHistory)
-	loadHistory(URLHist);
+	loadUrlHistory();
 #endif				/* not USE_HISTORY */
     known_hosts_file = rcFile("known_hosts");
     if (ssl_known_hosts)
@@ -2610,7 +2610,7 @@ _quitfm(int confirm)
 #endif				/* USE_COOKIE */
 #ifdef USE_HISTORY
     if (UseHistory && SaveURLHist)
-	saveHistory(URLHist, URLHistSize);
+	saveUrlHistory();
 #endif				/* USE_HISTORY */
     if ((err = save_known_hosts()))
 	perror(NULL);
